bokeh/models/mappers.py

```diff
@@ -24,6 +25,45 @@ class ColorMapper(Model):
     any of the palettes shown in :ref:`bokeh.palettes`.
     """).accepts(Enum(Palette), lambda pal: getattr(palettes, pal))
 
+    nan_color = Color(default="gray", help="""
+    Color to be used if data is NaN. Default: 'gray'
+    """)
+
+    def __init__(self, palette=None, **kwargs):
+        if palette is not None:
+            kwargs['palette'] = palette
+        super(ColorMapper, self).__init__(**kwargs)
+
+
+class CategoricalColorMapper(ColorMapper):
+    """ Map categories to colors. Values that are passed to
+    this mapper that aren't in factors will be assigned the nan_color.
+
+    """
+
+    factors = Either(Seq(String), Seq(Int), Seq(Float), Seq(Datetime), Seq(Date), help="""
+    A sequence of factors / categories that map to the color palette.
+    """)
+
+
+    def __init__(self, **kwargs):
+        super(ColorMapper, self).__init__(**kwargs)
+        palette = self.palette
+        factors = self.factors
+        if palette and factors:
+            if len(palette) != len(factors):
+                extra_factors = factors[len(palette):]
+                warnings.warn("""Palette length does not match number of
+factors. %s will be assigned to `nan_color` %s""" % (extra_factors, self.nan_color))
```

Comment:
This seems like an odd message if the palette is too long. `[] will be assigned to 'nan_color'`